File: internal/cliapp/timeago.go
package cliapp

import (
	"fmt"
	"time"
)

// timeAgo renders a stored timestamp relative to now, the way a person would
// say it: "just now", "4 min ago", "yesterday". A moment in the past is what
// the reader wants to place, and "2026-09-21T10:14:03.000Z" makes them do the
// arithmetic. Beyond a week the date itself is the clearer answer.
//
// A timestamp that cannot be parsed is shown as it is, rather than hidden: a
// wrong-looking value is a bug someone can report, and a blank is not.
func timeAgo(iso string, now time.Time) string {
	t, ok := parseSyncTime(iso)
	if !ok {
		return iso
	}
	d := now.Sub(t)
	switch {
	case d < 45*time.Second:
		// Covers a small clock skew between devices too: a stamp a few
		// seconds in the future is "just now", not a negative age.
		return "just now"
	case d < 90*time.Second:
		return "a minute ago"
	case d < time.Hour:
		return fmt.Sprintf("%d min ago", int(d.Minutes()+0.5))
	case d < 90*time.Minute:
		return "an hour ago"
	case d < 24*time.Hour:
		return fmt.Sprintf("%d hours ago", int(d.Hours()+0.5))
	case d < 36*time.Hour:
		return "yesterday"
	case d < 7*24*time.Hour:
		return fmt.Sprintf("%d days ago", int(d.Hours()/24+0.5))
	default:
		return "on " + t.Local().Format("2 Jan 2006")
	}
}

// parseSyncTime accepts both spellings a last-synced stamp has had: the
// millisecond form the daemon writes, and plain RFC 3339 from imports.
func parseSyncTime(iso string) (time.Time, bool) {
	if t, err := time.Parse("2006-01-02T15:04:05.000Z", iso); err == nil {
		return t, true
	}
	if t, err := time.Parse(time.RFC3339, iso); err == nil {
		return t, true
	}
	return time.Time{}, false
}

File: cmd/opensave-app/frontend/src/lib/timeago.js
// A stored moment, said the way a person would: "just now", "4 min ago",
// "yesterday". The same thresholds as the CLI's timeAgo, so the app and the
// terminal describe one sync the same way.
//
// Beyond a week the date itself is clearer than a count of days. Anything
// unparseable is returned as it came, so a wrong-looking value can be
// reported instead of vanishing.

const SEC = 1000;
const MIN = 60 * SEC;
const HOUR = 60 * MIN;
const DAY = 24 * HOUR;

export function timeAgo(iso, now = Date.now()) {
  if (!iso) return 'never';
  const t = Date.parse(iso);
  if (Number.isNaN(t)) return String(iso);
  const d = now - t;
  // A stamp a few seconds in the future is a peer's clock running slightly
  // ahead, not a negative age.
  if (d < 45 * SEC) return 'just now';
  if (d < 90 * SEC) return 'a minute ago';
  if (d < HOUR) return `${Math.round(d / MIN)} min ago`;
  if (d < 90 * MIN) return 'an hour ago';
  if (d < DAY) return `${Math.round(d / HOUR)} hours ago`;
  if (d < 36 * HOUR) return 'yesterday';
  if (d < 7 * DAY) return `${Math.round(d / DAY)} days ago`;
  return 'on ' + new Date(t).toLocaleDateString(undefined, { day: 'numeric', month: 'short', year: 'numeric' });
}

// The newest of several stamps, or null when there are none. Used where one
// game gets one line: "synced 4 min ago" is the most recent device.
export function latestOf(stamps) {
  let best = null;
  for (const iso of Object.values(stamps ?? {})) {
    if (iso && (best === null || iso > best)) best = iso;
  }
  return best;
}
